# 1 What is the Shell?

- The shell is a `program that takes keyboard commands and passes them to the operating system` to carry out. 
- Terminal emulators are `GUI for the shell`.
- If the last character of the prompt is a hash mark `(#)` rather than a dollar sign, the terminal session has `superuser` privileges.
- command history, linux distributions `remember the last 1,000 commands` by default
- double-clicking a word copies it

In [2]:
!date

Wed Nov 24 00:24:08 EST 2021


In [4]:
!cal

   November 2021      
Su Mo Tu We Th Fr Sa  
    1  2  3  4  5  6  
 7  8  9 10 11 12 13  
14 15 16 17 18 19 20  
21 22 23 24 25 26 27  
28 29 30              
                      


see current amount of free space on our disk

In [1]:
!df

Filesystem     1K-blocks      Used Available Use% Mounted on
/dev/sdd       263174212   3934396 245801660   2% /
tmpfs            2397672    410040   1987632  18% /mnt/wsl
/dev/sdc       263174212   7251308 242484748   3% /mnt/wsl/docker-desktop-data/isocache
none             2397672        12   2397660   1% /mnt/wsl/docker-desktop/shared-sockets/host-services
/dev/sdb       263174212    134176 249601880   1% /mnt/wsl/docker-desktop/docker-desktop-proxy
/dev/loop0        404596    404596         0 100% /mnt/wsl/docker-desktop/cli-tools
tools          498952516 217134732 281817784  44% /init
none             2395380         0   2395380   0% /dev
none             2397672         4   2397668   1% /run
none             2397672         0   2397672   0% /run/lock
none             2397672         0   2397672   0% /run/shm
none             2397672         0   2397672   0% /run/user
tmpfs            2397672         0   2397672   0% /sys/fs/cgroup
drivers        498952516 217134732 281817784  44

free memory

In [2]:
!free

              total        used        free      shared  buff/cache   available
Mem:        4795344     1145068     2623948      410328     1026328     3018808
Swap:       2097152        3596     2093556


# 2 Navigation

- `pwd` print name of current working directory
- `cd` change directory
- `ls` list directory contents

In [4]:
!pwd

/mnt/c/Users/yeman_s1h20q2/Yemane/Linux-Command-Line


In [6]:
!ls -a

.  ..  .git  LearningTheShell.ipynb  README.md


- absolute pathname starts from the root directory and leads to its destination
-a relative pathname starts from the working directory. 

- To do this, it uses `(dot) and .. (dot dot).` to represent relative positions in the file system tree. 

- The `. notation refers to the working directory`, and the `.. notation refers to the working directory’s parent directory.`

# 3 Exploring the System

- ls -a
    - show hidden files
- ls -l
    - output long format
- file 
    - determine file type
- less
    - view file contents

## Options and Arguments

Commands are often followed by one or more options.

```sh
command -options arguments
```

list files in long form (-l) and order by date modifcation(-t)

In [7]:
!ls -lt 

total 8
-rwxrwxrwx 1 yemane yemane 6150 Nov 24 18:34 LearningTheShell.ipynb
-rwxrwxrwx 1 yemane yemane   55 Nov 22 20:54 README.md


In [8]:
!ls -lt --reverse

total 8
-rwxrwxrwx 1 yemane yemane   55 Nov 22 20:54 README.md
-rwxrwxrwx 1 yemane yemane 6679 Nov 24 18:35 LearningTheShell.ipynb


## Long format

In [13]:
!ls -l

total 8
-rwxrwxrwx 1 yemane yemane 7197 Nov 24 18:38 LearningTheShell.ipynb
-rwxrwxrwx 1 yemane yemane   55 Nov 22 20:54 README.md


- access rights to a file [`-rwxrwxrwx`]
    - first character indicate file type
        - a dash (-) means regular file
        - a d means a directory
    - the next nine, split into groups of three characters are the access rights for the file owners, members of the files group, and everyone else respectively
- File number of hard links [1]
- username of the files owner [root]
- names of the hroup that owns the file [root]
- Size of file in bytes [7197]
- Date and itme of the files last modication [Nov 24 18:38]
- Name of file [LearningTheShell.ipynb]

In [14]:
!file LearningTheShell.ipynb

LearningTheShell.ipynb: UTF-8 Unicode text


In [16]:
!less README.md

# Linux-Command-Line
Learning the Linux Command Line
(END)

|Directory||
|-|-|
|/home|ordinary users can only write files only in their home directory|
|/lib|Contains shared library files used by the core system programs.|
|/lost+found|used in the case of a partial recovery from a file system corruption event|
|/media|contain the mount points for removable media such as USB drives|
|/mnt|contains mount points for removable devices that have been mounted manually|
|/opt|used to install “optional” software. This is mainly used to hold commercial software products|
|/proc|It’s not a real file system in the sense of files stored on your hard drive. Rather, it is a virtual file system maintained by the Linux kernel. The “files” it contains are peepholes into the kernel itself. The files are readable and will give you a picture of how the kernel sees your computer|
|/root|home directory for the root account|
|/sbin|This directory contains “system” binaries. These are programs that perform vital system tasks that are generally reserved for the superuser.|
|/tmp| intended for the storage of temporary, transient files created by various programs.|
|/usr|contains all the programs and support files used by regular user|
|/usr/bin|contains the executable programs installed by your Linux distribution|
|/usr/lib|The shared libraries for the programs in /usr/bin|
|/usr/local|where programs that are not included with your distribution but are intended for system-wide use are installed|
|/usr/sbin|Contains more system administration programs|
|/usr/share|contains all the shared data used by programs in /usr/bin|
|/usr/share/doc|we will find documentation files organized by packages|
|/var|where data that is likely to change is stored. Various databases, spool files, user mail, and so forth, are located here|
|/var/log| contains log files, records of various system activity.The most useful ones are /var/log/messages and /var/log/syslog|

## Symbolic Links

A way to reference a file with several names point to it.

# 4 Manipulating Files and Directories

|command||
|-|-|
|cp| copy files and directories|
|mv| move/rename files and directories|
|mkdir| create directories|
|rm| remove files and directories|
|ln| create hard and symbolic links|

## Wildcards

|wildcard|meaning|
|-|-|
|*|matches any character|
|?|matches any single character|
|[characters]|matches any character that is a member of the set characters|
|[!characters]|matches any character that not is a member of the set characters|
|[[:class:]]|matches any character that is a member of the specified class| 

|character class|meaning|
|-|-|
|[:alnum:]|matches any alphanumeric character|
|[:alpha:]|matches any alphabetic character|
|[:digit:]|matches any numeral|
|[:lower:]|matches any lowercase letter|
|[:upper:]|matches any uppercase letter\

In [2]:
!ls *

LearningTheShell.ipynb	README.md


In [5]:
!ls [[:upper:]]*.i*

LearningTheShell.ipynb


## Useful options

|options for cp||
|-|-|
|-a, --archive|copy file and directories and all their attributes including ownerships and permissions|
|-i, --interactive|before overwriting a file, prompt user for confirmation|
|-r, --recursive|recursively copy directories and their contents|
|-u, --update| when copying files from one directory into another, only copy files that either don't exist or  are newer than the existing corresponding files in the destination directory|
|-v, --verbose|display informative messages as the copt is performed|

|options for mv||
|-|-|
|-i, --interactive||
|-u, --update||
|-v, --verbose||

|options for rm||
|-|-|
|-i, --interactive||
|-r, --recursive||
|-f, --force|Ignore nonexistent files and do not prompt. This overrides the --interactive option.|
|-v, --verbose||

## ln - Create Links

The ln command is used to create either hard or symbolic links. It is used in one of two ways.

- hard link
    - ln file link
- symbolic link
    - ln -s item link

### Hard Links

Hard links are the original Unix way of creating links, compared to symbolic links, which are more modern. By default, every file has a single hard link that gives the file its name. When we create a hard link, we create an additional directory entry for a file

- A hard link cannot reference a file outside its own file system, disk partition
- A hard link may not reference a directory

A hard link is indistinguishable from the file itself.

### Symbolic Links
Work by creating a file that contains a text pointer to the referenced file or directory. In this way, they operate much in the same way of a `Windows shortcut`.

A file pointed to by a symbolic link and the symbolic link itself are 
largely indistinguishable from one another. For example, if you write something to the symbolic link, the referenced file is written to. When you delete a symbolic link, however, only the link is deleted, not the file itself. If the file is deleted before the symbolic link, the link will continue to exist but will point to nothing. In this case, the link is said to be broken. In many implementations, the ls command will display broken links in a distinguishing color, such as red, to reveal their presence

## Building a Playground

First create a directory to work in and call it playground.

In [33]:
mkdir playground

create a couple of directories inside it called dir1 and dir2

In [37]:
cd playground

/mnt/c/Users/yeman_s1h20q2/Yemane/Linux-Command-Line/playground


In [38]:
mkdir dir1 dir2

In [39]:
cp /etc/passwd .

In [40]:
ls -l

total 4
drwxrwxrwx 1 yemane yemane 4096 Nov 25 17:36 dir1/
drwxrwxrwx 1 yemane yemane 4096 Nov 25 17:36 dir2/
-rwxrwxrwx 1 yemane yemane 2321 Nov 25 17:39 passwd*


In [53]:
cp -v /etc/passwd .

'/etc/passwd' -> './passwd'


In [44]:
!file passwd

passwd: ASCII text


### Moving and Renaming Files

In [55]:
mv passwd fun

In [56]:
ls

dir1/  dir2/  fun*


In [57]:
mv fun dir1

In [58]:
ls dir1

fun*


In [59]:
mv dir1 dir2

In [61]:
ls -l dir2

total 0
drwxrwxrwx 1 yemane yemane 4096 Nov 25 17:47 dir1/


In [63]:
ls -l dir2/dir1/

total 4
-rwxrwxrwx 1 yemane yemane 2321 Nov 25 17:46 fun*


In [65]:
mv dir2/dir1 .

In [66]:
mv dir1/fun .

In [69]:
ls

dir1/  dir2/  fun*


### Creating hard links

In [71]:
!ln fun fun-hard

In [70]:
!ln fun dir1/fun-hard

In [72]:
!ln fun dir2/fun-hard

In [73]:
!ls -l

total 8
drwxrwxrwx 1 yemane yemane 4096 Nov 25 19:50 dir1
drwxrwxrwx 1 yemane yemane 4096 Nov 25 19:50 dir2
-rwxrwxrwx 4 yemane yemane 2321 Nov 25 17:46 fun
-rwxrwxrwx 4 yemane yemane 2321 Nov 25 17:46 fun-hard


both the second fields in the listings for fun and fun-hard contain a 4, which is the number of hard links that now exist for the file. Remember that a file will always have at least one link because the file’s name is created by a link.

When thinking about hard links, it is helpful to imagine that files are made up of two parts.
    - the data part containing the file's contents
    - the name part that holds the file's name

When we created hard links, we are actually creating additional name parts that all refer to the same data part. 

In [76]:
!ls -li

total 8
 12947848929171375 drwxrwxrwx 1 yemane yemane 4096 Nov 25 19:50 dir1
 12666373952460722 drwxrwxrwx 1 yemane yemane 4096 Nov 25 19:50 dir2
116530640358274685 -rwxrwxrwx 4 yemane yemane 2321 Nov 25 17:46 fun
116530640358274685 -rwxrwxrwx 4 yemane yemane 2321 Nov 25 17:46 fun-hard


the first field is the inode number and, as we can see, both fun and fun-hard share the same inode number, which confirms they are the same file

### Creating symbolic links

Symbolic links were created to overcome the two disadvantages of hard links:
    - hard links cannot span physical devices
    - hard links cannot reference directories, only files

Symbolic links are a special type of file that contains a text pointer to the target file or directory.

In [77]:
!ln -s fun fun-smy

In [79]:
!ln -s ../fun dir1/fun-sym

In [80]:
!ln -s ../fun dir2/fun-sym

The listing for fun-sym in dir1 shows that it is a symbolic link by the leading l in the first field and that it points to ../fun, which is correct. Relative to the location of fun-sym, fun is in the directory above it. Notice, too, that the length of the symbolic link file is 6, the number of characters in the string ../fun rather than the length of the file to which it is pointing.

In [81]:
!ls -l dir1

total 4
-rwxrwxrwx 4 yemane yemane 2321 Nov 25 17:46 fun-hard
lrwxrwxrwx 1 yemane yemane    6 Nov 25 20:00 fun-sym -> ../fun


### Removing files and Directories

In [82]:
!rm fun-hard

That worked as expected. The file fun-hard is gone, and the link count 
shown for fun is reduced from four to three, as indicated in the second field of the directory listing.

In [83]:
!ls -l

total 4
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir1
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir2
-rwxrwxrwx 3 yemane yemane 2321 Nov 25 17:46 fun
lrwxrwxrwx 1 yemane yemane    3 Nov 25 20:00 fun-smy -> fun


In [84]:
!rm fun

In [85]:
!ls -l

total 0
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir1
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir2
lrwxrwxrwx 1 yemane yemane    3 Nov 25 20:00 fun-smy -> fun


In [86]:
!less fun-smy

fun-smy: No such file or directory


In [89]:
!rm fun-smy 

rm: cannot remove 'fun-smy': No such file or directory


In [90]:
!ls -l

total 0
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir1
drwxrwxrwx 1 yemane yemane 4096 Nov 25 20:00 dir2


In [105]:
cd ..

/mnt/c/Users/yeman_s1h20q2/Yemane/Linux-Command-Line


In [108]:
!rm -r playground

In [109]:
!ls

LearningTheShell.ipynb	README.md
